### ISSUE TITLE
Cybersecurity Intrusion Detection with Machine Learning

### ISSUE THREAD

**Issue by akinyeraakintunde** [akinyeraakintunde/Porfolio-Project#1]
This project demonstrates how machine‑learning techniques can be applied to a simplified intrusion detection problem.
By generating a synthetic network‑traffic dataset and training a random‑forest classifier, the project illustrates end‑to‑end data science skills—from data generation and preprocessing to model development, evaluation and reporting.
The repository can be published publicly to showcase technical capability and innovation as part of a UK Tech Nation Global Talent application.

About the Dataset

The dataset is created using scikit‑learn’s make_classification function. It contains 1,000 samples with 10 features representing network characteristics such as connection duration, protocol type, source and destination bytes, flags and login attempts.
The target variable labels each sample as benign traffic (0) or an attack (1). The class distribution is intentionally imbalanced (approximately 70:30) to reflect real‑world conditions.

While synthetic, the dataset structure is inspired by the NSL‑KDD intrusion detection benchmark. Future iterations could substitute this with the actual NSL‑KDD dataset or other open datasets to validate the approach on real data.

Repository Contents
FileDescriptionnetwork_intrusion_detection.pyPython script that generates the synthetic dataset, trains Random Forest and Gradient Boosting classifiers, evaluates their performance, prints metrics and saves visualisation images (confusion matrices and feature importances).requirements.txtList of Python dependencies required to run the script (scikit‑learn, pandas, numpy).
Running the Project
1. Set up the environment. Create and activate a virtual environment (optional) and install dependencies:
python -m venv venv
source venv/bin/activate
pip install -r requirements.txt
1. 
2. Run the script. Execute the Python script to generate data, train the model and display results:
python network_intrusion_detection.py
1. 
2. Interpret the output. The script prints accuracy, precision, recall and F1‑scores for each model and generates PNG images in an outputs/ directory: confusion matrices for both models and a feature‑importance bar chart for the Random Forest. In initial experiments, both models achieved accuracy above 98 %, with very few false positives or negatives.

Results Summary

On the synthetic dataset, the random‑forest classifier achieved the following performance:
– Accuracy: >98 %
– Precision (benign vs. attack): 0.98 / 1.00
– Recall: 1.00 / 0.96
– F1‑score: 0.99 / 0.98

The confusion matrix shows very few false positives and false negatives, suggesting the model is effective at distinguishing attacks from normal traffic. Feature importances (available via the model’s feature_importances_ attribute) reveal which network characteristics have the greatest influence on predictions.

Future Work
– Replace the synthetic dataset with a real intrusion detection dataset such as NSL‑KDD or CICIDS2017 to validate the model in realistic scenarios.
– Experiment with other algorithms (e.g., XGBoost, LightGBM, deep neural networks) and compare their performance.
– Implement feature engineering techniques to capture temporal patterns and complex relationships in network traffic.
– Deploy the trained model as an API or integrate it into a SIEM (Security Information and Event Management) system for real‑time monitoring.

License

This project is released under the MIT License. You are free to use, modify and distribute it for any purpose.
